File: rhinocommon/snippets/cs/pre-and-post-pick-objects.cs
partial class Examples
{
  public static Result PrePostPick(RhinoDoc doc)
  {
    var go = new Rhino.Input.Custom.GetObject();
    go.SetCommandPrompt("Select objects");
    go.EnablePreSelect(true, true);
    go.EnablePostSelect(true);
    go.GetMultiple(0, 0);
    if (go.CommandResult() != Result.Success)
      return go.CommandResult();

    var selected_objects = go.Objects().ToList();

    if (go.ObjectsWerePreselected)
    {
      go.EnablePreSelect(false, true);
      go.DeselectAllBeforePostSelect = false;
      go.EnableUnselectObjectsOnExit(false);
      go.GetMultiple(0, 0);
      if (go.CommandResult() == Result.Success)
      {
        foreach (var obj in go.Objects())
        {
          selected_objects.Add(obj);
          // The normal behavior of commands is that when they finish,
          // objects that were pre-selected remain selected and objects
          // that were post-selected will not be selected. Because we
          // potentially could have both, we'll try to do something
          // consistent and make sure post-selected objects also stay selected
          obj.Object().Select(true);
        }
      }
    }
    return selected_objects.Count > 0 ? Result.Success : Result.Nothing;
  }
}

File: rhinocommon/snippets/py/pre-and-post-pick-objects.py
from Rhino import *
from Rhino.Commands import *
from Rhino.Input.Custom import *

def RunCommand():
  go = GetObject()
  go.SetCommandPrompt("Select objects")
  go.EnablePreSelect(True, True)
  go.EnablePostSelect(True)
  go.GetMultiple(0, 0)
  if go.CommandResult() != Result.Success:
    return go.CommandResult()

  selected_objects = []
  for obj in go.Objects():
    selected_objects.Add(obj)

  if go.ObjectsWerePreselected:
    go.EnablePreSelect(False, True)
    go.DeselectAllBeforePostSelect = False
    go.EnableUnselectObjectsOnExit(False)
    go.GetMultiple(0, 0)
    if go.CommandResult() == Result.Success:
      for obj in go.Objects():
        selected_objects.Add(obj)
        # The normal behavior of commands is that when they finish,
        # objects that were pre-selected remain selected and objects
        # that were post-selected will not be selected. Because we
        # potentially could have both, we'll try to do something
        # consistent and make sure post-selected objects also stay selected
        obj.Object().Select(True)
  return Result.Success if selected_objects.Count > 0 else Result.Nothing

if __name__ == "__main__":
  RunCommand()
